worker/caasfirewallerembedded/applicationworker.go

```diff
@@ -0,0 +1,262 @@
+	"reflect"
+	"strings"
+
+	"github.com/juju/charm/v8"
+	"github.com/juju/errors"
+	"github.com/juju/worker/v2"
+	"github.com/juju/worker/v2/catacomb"
+
+	"github.com/juju/juju/caas"
+	"github.com/juju/juju/core/network"
+	"github.com/juju/juju/core/watcher"
+)
+
+type applicationWorker struct {
+	catacomb       catacomb.Catacomb
+	controllerUUID string
+	modelUUID      string
+	appName        string
+
+	firewallerAPI CAASFirewallerAPI
+
+	broker         CAASBroker
+	portMutator    PortMutator
+	serviceUpdater ServiceUpdater
+
+	appWatcher   watcher.NotifyWatcher
+	portsWatcher watcher.StringsWatcher
+
+	lifeGetter LifeGetter
+
+	initial           bool
+	previouslyExposed bool
+
+	currentPorts portRanges
+
+	logger Logger
+}
+
+type portRanges map[network.PortRange]bool
+
+func newPortRanges(in []network.PortRange) portRanges {
+	out := make(portRanges)
+	for _, p := range in {
+		out[p] = true
+	}
+	return out
+}
+
+func (pg portRanges) equal(in portRanges) bool {
+	if len(pg) != len(in) {
+		return false
+	}
+	return reflect.DeepEqual(pg, in)
+}
+
+func (pg portRanges) toServicePorts() []caas.ServicePort {
+	out := make([]caas.ServicePort, len(pg))
+	for p := range pg {
+		out = append(out, caas.ServicePort{
+			// TODO(embedded): add name to `network.PortRange`?
+			// Name:       p.Name,
+			Port:       p.FromPort,
+			TargetPort: p.ToPort,
+			Protocol:   p.Protocol,
+		})
+	}
+	return out
+}
+
+func newApplicationWorker(
+	controllerUUID string,
+	modelUUID string,
+	appName string,
+	firewallerAPI CAASFirewallerAPI,
+	broker CAASBroker,
+	lifeGetter LifeGetter,
+	logger Logger,
+) (worker.Worker, error) {
+	w := &applicationWorker{
+		controllerUUID: controllerUUID,
+		modelUUID:      modelUUID,
+		appName:        appName,
+		firewallerAPI:  firewallerAPI,
+		broker:         broker,
+		lifeGetter:     lifeGetter,
+		initial:        true,
+		logger:         logger,
+	}
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+	}); err != nil {
+		return nil, errors.Trace(err)
+	}
+	return w, nil
+}
+
+// Kill is part of the worker.Worker interface.
+func (w *applicationWorker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *applicationWorker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+func (w *applicationWorker) setUp() (err error) {
+	w.appWatcher, err = w.firewallerAPI.WatchApplication(w.appName)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err := w.catacomb.Add(w.appWatcher); err != nil {
+		return errors.Trace(err)
+	}
+
+	w.portsWatcher, err = w.firewallerAPI.WatchOpenedPorts()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err := w.catacomb.Add(w.portsWatcher); err != nil {
+		return errors.Trace(err)
+	}
+
+	charmURL, err := w.firewallerAPI.ApplicationCharmURL(w.appName)
+	if err != nil {
+		return errors.Annotatef(err, "failed to get charm urls for application")
+	}
+	charmInfo, err := w.firewallerAPI.CharmInfo(charmURL.String())
```

Comment:
Instead of 2 api calls - one to get the url then another to get the metadata, can we just get the metadata with one api call